source/SH3/graphics/texture.cpp

```diff
@@ -13,48 +13,330 @@
 #include <SH3/arc/vfile.hpp>
 
 #include <cassert>
+#include <fstream>
 
+#define HEADER_START 64
+
+static std::size_t offset = 0;
 
 using namespace sh3_graphics;
 
+#pragma pack(push, 1)
+struct rgba
+{
+    std::uint8_t r, g, b, a;
+};
+
+struct rgba16
+{
+    uint8_t r : 5;
+    uint8_t g : 5;
+    uint8_t b : 5;
+    uint8_t a : 1;
+};
+
+struct rgb
+{
+    std::uint8_t r, g, b;
+};
+
+struct bgra
+{
+    std::uint8_t b, g, r, a;
+};
+
+struct bgr
+{
+    std::uint8_t b, g, r;
+};
+
+struct tga_header
+{
+    std::uint8_t id_size;   // Size of the ID field that follows this header (usually 0)
+    std::uint8_t colormap;  // Is this image paletted (THIS MUST ALWAYS BE 0 FOR US!)
+    std::uint8_t type;      // Type of image (ALWAYS 2 FOR RGB24!)
+    std::uint8_t unused[5]; // Colormap crap
+
+    std::uint16_t x_origin;  // Co-ordinate for first X value
+    std::uint16_t y_origin;  // Co-ordinate for first Y value
+    std::uint16_t width;
+    std::uint16_t height;
+    std::uint8_t bpp;       // Bits per pixel
+    std::uint8_t flags;
+};
+
+#pragma pack(pop)
+
+/**
+ *  Dump a texture to a Microsoft BMP/DIB
+ *
+ *  @param width - The Width of this texture.
+ *  @param height - The Height of this texture.
+ *  @param data - Pixel data
+ */
+void Dump2Bitmap(std::uint32_t width, std::uint32_t height, std::vector<std::uint8_t>& data)
+{
+    tga_header header;
+    std::ofstream file("output.tga", std::ios::binary); // Open the stream for binary output
+
+    memset(&header, 0, sizeof(header));
+
+    header.id_size      = 0x00;
+    header.colormap     = 0x00;
+    header.type         = 2;
+
+    header.x_origin     = 0x0000;
+    header.y_origin     = 0x0000;
+    header.width        = (uint16_t)width;
+    header.height       = (uint16_t)height;
+    header.bpp          = 24;
+    header.flags        = 0x20; // Origin is upper left hand corner
+
+    file.write((char*)&header, sizeof(header));
+    file.write((char*)&data[0], data.size());
+}
+
 //TODO: Scale the texture and then
 void sh3_texture::Load(sh3_arc& mft, const std::string& filename)
 {
     sh3_texture_header          header;
     sh3_arc_vfile               file(mft, filename);
     sh3_arc_vfile::read_error   e;
-    std::vector<std::uint8_t>   data;     // Pixel data of this texture (with the header stripped)
+    std::vector<uint8_t>        data;       // Pixel data of this texture (with the header stripped)
+    std::vector<rgba>           palette;    // Palette Data (I think this is BGRA)
+    std::size_t                 ret;        // How many bytes we get in from file.ReadData( )
 
-    std::size_t ret = file.ReadData(&header, sizeof(header), e); // Read in the header
```

Comment:
I think it made sense to declare `ret` here. This removes the need do document what we use `ret` for above.

In general I'm in favor of the "declare where used paradigm".